cpu-o3: Transform the lsqunit

happy-lx opened the pull request:
## Background
In order to align the behavior of the LSU module with the RTL, this PR makes the following modifications:
+ Uses TimeBuffer to visualize the load and store pipelines.
+ Distributes the operational details of the load and store pipelines across the corresponding pipeline stages.
+ Adds nuke replay operation to reduce RAW violations.
+ Aligns miss load replay behavior with RTL and adds early wake-up.

### 1. Pipeline Construction
Related commits:
+ [f5039874530fdd4ec7ca8097e13c8c3d20607e84](https://github.com/OpenXiangShan/GEM5/pull/265/commits/f5039874530fdd4ec7ca8097e13c8c3d20607e84)
+ [89fdfd0a589d5390e12d2c016fa99d6b59f4e1c1](https://github.com/OpenXiangShan/GEM5/pull/265/commits/89fdfd0a589d5390e12d2c016fa99d6b59f4e1c1)
+ [bfbf7c957c2a873aee9105664d1228233ceb4f92](https://github.com/OpenXiangShan/GEM5/pull/265/commits/bfbf7c957c2a873aee9105664d1228233ceb4f92)

**Original design:** Load/store operations are delayed by a certain number of cycles and then all operations (TLB, Cache lookup, exception check, etc.) are completed in one cycle using `executeLoad/executeStore`.

**New design:** TimeBuffer is set according to the corresponding pipeline stages. Instructions are first dispatched to TimeBuffer `s0`, and over time they pass through the stages of the pipeline to complete the corresponding operations.

<details>
  <summary>Code Details</summary>

  **Original design:**

  + The operation was delayed by `op_latency - 1` cycles before being sent to the corresponding function unit. https://github.com/OpenXiangShan/GEM5/blob/2d1995aa9820b5b26b68002e698960b628e072cb/src/cpu/o3/inst_queue.cc#L700-L708
  + After reaching the function unit, the load/store calls `executeLoad/executeStore` to complete the corresponding operations. https://github.com/OpenXiangShan/GEM5/blob/2d1995aa9820b5b26b68002e698960b628e072cb/src/cpu/o3/iew.cc#L1407-L1437
  + `executeLoad/executeStore` completes all operations. https://github.com/OpenXiangShan/GEM5/blob/2d1995aa9820b5b26b68002e698960b628e072cb/src/cpu/o3/lsq_unit.cc#L906-L979 https://github.com/OpenXiangShan/GEM5/blob/2d1995aa9820b5b26b68002e698960b628e072cb/src/cpu/o3/lsq_unit.cc#L1003-L1067

   **New design:**

  + The load/store operations are directly sent to the function unit. https://github.com/OpenXiangShan/GEM5/blob/cf1302416f681eded0b2ba9f229dc27434cdfb83/src/cpu/o3/inst_queue.cc#L711-L719
  + After reaching the function unit, load/store operations call `issueToLoadPipe/issueToStorePipe` to send the instructions to the `s0` of the load/store pipeline. `loadPipeSx`/`storePipeSx` is the corresponding pipeline timebuffer, `loadPipeSx[0]` means s0 stage of load pipeline.  https://github.com/OpenXiangShan/GEM5/blob/cf1302416f681eded0b2ba9f229dc27434cdfb83/src/cpu/o3/lsq_unit.hh#L658-L682 https://github.com/OpenXiangShan/GEM5/blob/cf1302416f681eded0b2ba9f229dc27434cdfb83/src/cpu/o3/iew.cc#L1473-L1503 https://github.com/OpenXiangShan/GEM5/blob/cf1302416f681eded0b2ba9f229dc27434cdfb83/src/cpu/o3/lsq_unit.cc#L1053-L1081
  + Over time, the instructions gradually enter the subsequent pipeline stages (`s1`, `s2`, etc.). The corresponding instructions are fetched from the load/store pipeline at each stage and the operations are executed. https://github.com/OpenXiangShan/GEM5/blob/cf1302416f681eded0b2ba9f229dc27434cdfb83/src/cpu/o3/lsq_unit.cc#L1314-L1367 https://github.com/OpenXiangShan/GEM5/blob/cf1302416f681eded0b2ba9f229dc27434cdfb83/src/cpu/o3/lsq_unit.cc#L1493-L1559
</details>

### 2. Nuke Replay
Related commit:
+ [bfbf7c957c2a873aee9105664d1228233ceb4f92](https://github.com/OpenXiangShan/GEM5/pull/265/commits/bfbf7c957c2a873aee9105664d1228233ceb4f92)

> Scenario:
> ```asm
> sw s1, 0(s0)
> lw s2, 0(s0)
> ```
> The load needs to forward data from the store, and both load and store are issued simultaneously from the IQ into the pipeline, arriving at `s1` in the load/store pipeline at the same time.

**Original design:** If the load executes first, since the store in the same cycle hasn't finished (and hasn't updated the SQ), the load can't detect the dependency and thus can't forward data from the store. Then the store in the same cycle executes, a RAW violation will be detected, causing the pipeline to be flushed.

**What about RTL behavior:** The described situation does not occur in the RTL because the store instruction in the s1 pipeline stage (the cycle before the SQ is updated) checks whether there is a load instruction on the load pipeline with the same address and data that has not forwarded from it, and causes that load instruction to be replayed.

**New design:** When the load reaches `s1` or `s2`, it checks whether there are stores still in `s1` that haven't executed and match the address. If so, the load will be replayed. This is called the pipeline nuke replay. After the store in `s1` completes, the address and data are updated to the SQ, allowing the replayed load to correctly forward data without causing a RAW violation.

> **You can use `EnablePipeNukeCheck` to control whether to enable this feature. In addition, if you want to use this feature, make sure `EnableLdMissReplay` is set to `True`.**

<details>
  <summary>Code Details</summary>

**RTL design:**
  + store instructions at store pipeline `s1` query load pipeline `s1/s2`, causing the matched load to be replayed https://github.com/OpenXiangShan/XiangShan/blob/0051450372ae5a03ce9d36afdbdd34b9a19f4785/src/main/scala/xiangshan/mem/pipeline/LoadUnit.scala#L971-L995 https://github.com/OpenXiangShan/XiangShan/blob/0051450372ae5a03ce9d36afdbdd34b9a19f4785/src/main/scala/xiangshan/mem/pipeline/LoadUnit.scala#L1228-L1248 https://github.com/OpenXiangShan/XiangShan/blob/0051450372ae5a03ce9d36afdbdd34b9a19f4785/src/main/scala/xiangshan/mem/pipeline/LoadUnit.scala#L1357-L1381

 **Gem5 New design:**

  + The `pipeLineNukeCheck` is where the load in `s1` checks for pipeline nuke. https://github.com/OpenXiangShan/GEM5/blob/cf1302416f681eded0b2ba9f229dc27434cdfb83/src/cpu/o3/lsq_unit.cc#L1159-L1179
  + The `pipeLineNukeCheck` for load in `s2` checks for pipeline nuke. https://github.com/OpenXiangShan/GEM5/blob/cf1302416f681eded0b2ba9f229dc27434cdfb83/src/cpu/o3/lsq_unit.cc#L1250-L1258
  + Check logic. https://github.com/OpenXiangShan/GEM5/blob/cf1302416f681eded0b2ba9f229dc27434cdfb83/src/cpu/o3/lsq_unit.cc#L743-L762
  + If the load in `s1` or `s2` triggers a pipeline nuke, a fast load replay is performed in `s2`. https://github.com/OpenXiangShan/GEM5/blob/cf1302416f681eded0b2ba9f229dc27434cdfb83/src/cpu/o3/lsq_unit.cc#L1273-L1282
  + If the load needs to perform a pipeline nuke replay, the pipeline flag of this load will be marked as `Nuke`, and the RAW violation check will skip this load. https://github.com/OpenXiangShan/GEM5/blob/cf1302416f681eded0b2ba9f229dc27434cdfb83/src/cpu/o3/lsq_unit.cc#L955-L961
</details>

### 3. Miss Load Replay  
Commits related to this modification:  
+ [ee53d22929c24e56c386efc13190b74c90f8997c](https://github.com/OpenXiangShan/GEM5/pull/265/commits/ee53d22929c24e56c386efc13190b74c90f8997c)  
+ [c3cd30088a4d5a57a47b54a43c53473436b5650c](https://github.com/OpenXiangShan/GEM5/pull/265/commits/c3cd30088a4d5a57a47b54a43c53473436b5650c)  

**Original design:** A miss load directly writes back after receiving the `TimingResp`, with no limit on the number and no re-entering pipeline behavior.

**New design:** A miss load receives a custom `Hint` signal two cycles before receiving the `TimingResp`, which wakes up the relevant instructions and reissues them into the pipeline. The relevant data becomes stable in the `bus` after `TimingResp` reaching the LSU, and the reissued load will forward data from the `bus`. If data cannot be forwarded, it will access the cache again. After the cache refilled the data to itself completely, a `Bus_Clear` request is sent to the LSU to clear the corresponding data in the `bus`. 

> **You can use `EnableLdMissReplay` to control whether to enable this feature.**

<details>  
  <summary>Code details</summary>  

  **Original design:**

  + A miss load directly calls `completeDataAccess` for write-back after receiving the `TimingResp` https://github.com/OpenXiangShan/GEM5/blob/2d1995aa9820b5b26b68002e698960b628e072cb/src/cpu/o3/lsq.cc#L1349-L1362

  **New design:**

  + A miss load enters a `cache miss load replayQueue` https://github.com/OpenXiangShan/GEM5/blob/cf1302416f681eded0b2ba9f229dc27434cdfb83/src/cpu/o3/lsq_unit.cc#L1259-L1271
  + When the refill data reaches the bus between L2 and L1, a `Hint` signal is returned a few cycles ahead of the `TimingResp` https://github.com/OpenXiangShan/GEM5/blob/cf1302416f681eded0b2ba9f229dc27434cdfb83/src/mem/coherent_xbar.cc#L518-L522
  + After the `Hint` reaches L1, it queries the MSHR and sends a wake-up signal to the relevant load requests in the cache block. The loads that receive the wake-up signal are awakened from the replayQueue https://github.com/OpenXiangShan/GEM5/blob/cf1302416f681eded0b2ba9f229dc27434cdfb83/src/mem/cache/cache.cc#L963-L999 https://github.com/OpenXiangShan/GEM5/blob/cf1302416f681eded0b2ba9f229dc27434cdfb83/src/cpu/o3/lsq.cc#L1524-L1537
  + The actual `TimingResp` arrives a few cycles after the `Hint` and stabilizes the data on the bus https://github.com/OpenXiangShan/GEM5/blob/cf1302416f681eded0b2ba9f229dc27434cdfb83/src/cpu/o3/lsq.cc#L1434-L1475
  + The reissued load checks the bus before accessing the cache. If data is available on the bus, it directly uses the data from the bus (via `forwardFrmBus(load_inst, request)`), and accesses the cache only if no data is found on the bus. https://github.com/OpenXiangShan/GEM5/blob/cf1302416f681eded0b2ba9f229dc27434cdfb83/src/cpu/o3/lsq_unit.cc#L2995-L3023
  + Sure, it is also possible that the Hint is sent too early, causing the reissued load reaching the pipeline and find that the TimingResp has not yet arrived. In this case, the load will speculatively wait for one more cycle, i.e., wait until s2. If the TimingResp arrives, it can forward data from the bus in s2 and then write back. If s2 still hasn't received the TimingResp, it will re-enter the cache miss replay queue.(**Note: this scenario will be very rare.**) https://github.com/OpenXiangShan/GEM5/blob/cf1302416f681eded0b2ba9f229dc27434cdfb83/src/cpu/o3/lsq_unit.cc#L2995-L3002 https://github.com/OpenXiangShan/GEM5/blob/cf1302416f681eded0b2ba9f229dc27434cdfb83/src/cpu/o3/lsq_unit.cc#L1221-L1244
  + After the cache refill data is ready within the cache, a `Bus_Clear` is sent to the LSU to clear the relevant data from the bus https://github.com/OpenXiangShan/GEM5/blob/cf1302416f681eded0b2ba9f229dc27434cdfb83/src/mem/cache/base.cc#L2041-L2052 https://github.com/OpenXiangShan/GEM5/blob/cf1302416f681eded0b2ba9f229dc27434cdfb83/src/cpu/o3/lsq.cc#L586-L602
</details>

### 4. Misc
During the alignment process, some functionality correctness issues and minor misalignments were also discovered, and they are all documented in this section.

#### 4.1 fence opType
Commits related to this modification:  
+ [5e118d7bfbffc7d18a4ad52b73330e61221e7713](https://github.com/OpenXiangShan/GEM5/pull/265/commits/5e118d7bfbffc7d18a4ad52b73330e61221e7713)  

**Original design:**

the fence instruction will be dispatched to mem's dispatchQueue, but its opType is `No_OpClass`, which will cause it to wait for the integer issue queue IQ2(IntMisc) to have free items before it can continue execution.
If the subsequent instructions of the fence instruction occupy the intIQ2, the fence cannot be executed and cpu deadlocks.

**new design:** 

Change the opType of the fence instruction to `MemReadOp` to prevent this situation (in fact, the fence will not be dispatched to IQ)

#### 4.2 LRSC
Commits related to this modification:  
+ [af375fa600564b38392501242e9aa1ddd19d2b56](https://github.com/OpenXiangShan/GEM5/pull/265/commits/af375fa600564b38392501242e9aa1ddd19d2b56)  

LR Instruction can be executed speculatively, which causes RAW violations in some corner cases.
**To reduce complexity and ensure consistency with the RTL design, avoid speculative execution for the LR, and strictly maintain order.**

#### 4.3 store writeback
Commits related to this modification:  
+ [3891ceb1bcc9cb2245b02c61e1fad3ecbefc6758](https://github.com/OpenXiangShan/GEM5/pull/265/commits/3891ceb1bcc9cb2245b02c61e1fad3ecbefc6758)

Previously, the store was written back in s2, but it is written back in s4 in RTL design. So align the timing of this operation. **The write-back timing has a significant impact on high IPC programs like hmmer (~13%), with earlier write-back being more beneficial for improving performance.**

> **You can set `StoreWbStage` to 2 or 4 or N to control which stage the store is written back.**

happy-lx committed:
cpu-o3: replay cache missed load from replayQ

This commit is only for normal load. The uncache/amo load is the same as the original process.

Change-Id: Idc98ee18a6e94a39774ebba0f772820699b834de

happy-lx committed:
cpu-o3: Transform the lsqunit

Transform the load/store execution logic into a multi-stage pipeline form

Change-Id: Iaf7558ad75ed8fe2bbf4a776359db113b6126453

happy-lx committed:
arch-riscv: change fence's opType

Originally, the fence instruction will be dispatched to mem's dispatchQueue,
but its opType is No_OpClass, which will cause it to wait for the integer issue queue IQ2(IntMisc) to have free items before it can continue execution.

If the subsequent instructions of the fence instruction occupy the intIQ2, the fence cannot be executed and cpu stucks.

Therefore, change the opType of the fence instruction to MemReadOp to prevent this situation (in fact, the fence will not be dispatched to IQ)

Change-Id: Ie38a901e038db9906c43f78675e69391e847c88b

happy-lx committed:
cpu-o3: Split operations in the ldst pipeline

Now initiateAcc only does tlb access and is located at s0 of the load/store pipeline.

Load makes cache access and query violations at s1, receives the cache response at s2, and writes back at s3.
Store updates sq and query violations at s1, and writes back at s4.

AMO operations are now executed using `executeAmo`.

Change-Id: Iac678b7de3a690329f279c70fdcd22be4ed22715

happy-lx committed:
arch: use strictly order-preserving LRSC

Add a fence before and after the LRSC instruction.

Change-Id: I66021d0a5a653d2a7e30cd262166363a84184ed6

happy-lx committed:
mem: let load has certain latency in ruby cahche

Change-Id: Ifc1a586df8beab65772d48a75106155f9e723cba

happy-lx committed:
cpu-o3: tune the behavior of the ldst pipeline

Adjust cache miss load replay logic: replay all loads cannot get data at
load s2, now we don’t need cache to send `sendCustomSignal` when miss.

Add RAW nuke replay at load s1&s2

Move most of the writeback logic to load s2 and actually writeback at s3

Change-Id: Idfd3480969958826f4820349168f17c9522f791e

happy-lx committed:
cpu-o3: add params to control nuke and miss replay

set `EnableLdMissReplay` to True to enable replaying missed load from
replayQ

set `EnablePipeNukeCheck` to True to detect raw nuke replay in loadpipe

NOTE: if `Enableldmissreplay` is False, `EnablePipeNukeCheck` can't be
set as True

Change-Id: Ic4235bffba01d5dc4c39cec8ae92f2d27b28d98a

happy-lx committed:
cpu-o3: make store wb stage configurable

store writeback at S4 by default
when using --ideal-kmhv3, store writeback at S2

Change-Id: I6a318ff6c182daca0ab041840d76575a16e45d82

happy-lx committed:
cpu-o3: refactor fullforward code

Change-Id: I5829589df8ca01724ffa4369d23d7e4693e0aea1

happy-lx committed:
mem: fix write packet latency calculation

Previously, the delay of the write packet operation did not take into account
whether the block was ready. In fact, if the block is not ready,
the actual timing for the write to return the TimingResp should be delayed.

Change-Id: I65de8d47e2f24ad4be867e1867cddee06092f22f

happy-lx committed:
mem: add Load Custom Hint Wakeup

Currently, at the xbar, except sending the actual TimingResp, a Hint signal is sent N cycles in advance.
(N is set by hint_wakeup_ahead_cycles in Caches.py)
This Hint signal first queries the MSHR, finds all the associated load instructions, and issues a custom wake-up.
Once all the custom wake-ups are received by the load instructions, they wake up the load instructions in the replayQueue.
When these waken instructions reach stage s1 or s2 at load pipeline, data is forwarded from the bus.

The actual TimingResp will place the data on the bus until the Dcache finally writes the data to itself then clears it.

Change-Id: I8960acc14e95c06d8b1a86220f36a181588ff7f4

happy-lx committed:
cpu-o3: delete pending wake events if cancel

If this instruction is cancelled, the associated wake event should be descheduled.

Change-Id: I595541aa5f96163350aa5f6e3825f78520a0e660

tastynoob commented on src/cpu/o3/lsq.cc:
```diff
@@ -121,9 +135,10 @@ LSQ::LSQ(CPU *cpu_ptr, IEW *iew_ptr, const BaseO3CPUParams &params)
     }
 
     thread.reserve(numThreads);
+    // TODO: Parameterize the load/store pipeline stages
```
parameterize this

happy-lx replied:
OK, I will fix this

tastynoob commented on src/cpu/o3/lsq.cc:
```diff
@@ -1282,7 +1365,9 @@ LSQ::LSQRequest::~LSQRequest()
     }
     assert(!isAnyOutstandingRequest());
-    _inst->savedRequest = nullptr;
+    if (_inst->savedRequest == this) {
```
why need this condition

happy-lx replied:
In the current design, after a load instruction experiences a miss for the first time, it does not continue to track the corresponding request. 

Once the miss request retrieves the data, refills the dcache, and returns a response to the LSU, it will discard itself. 

When the missed load is replayed to the pipeline, a new request will be generated(the goal is to ensure it re-accesses the DTLB and dcache when necessary). **This approach indeed breaks the rule that each instruction is correspond to only one request. If the request management is not handled properly, it could potentially lead to functional issues.**

However, in order to implement early wake-up and align with the RTL, I feel this approach is actually more friendly in terms of implementation.

tastynoob commented on src/cpu/o3/lsq_unit.cc:
```diff
@@ -947,35 +1156,413 @@ LSQUnit::executeLoad(const DynInstPtr &inst)
+}
+
+Fault
+LSQUnit::storePipeS2(const DynInstPtr &inst, std::bitset<LdStFlagNum> &flag)
```
Why so many identical function?
Can be simplified

happy-lx replied:
Yes. These `storePipeS2`, `S3`, and `S4` do not perform any operations and are indeed redundant. I will replace them with a new function.

tastynoob commented on src/cpu/o3/lsq_unit.cc:
```diff
@@ -1973,6 +2548,40 @@ LSQUnit::recvRetry()
+void
+LSQUnit::dumpLoadPipe()
+{
+    DPRINTF(LSQUnit, "Dumping LoadPipe:\n");
```
use:
if (debug::LSQUnit)

happy-lx replied:
Yes, using `debug::LSQUnit` is much better for simulation speed, I will fix it.

tastynoob commented on src/cpu/o3/lsq_unit.cc:
```diff
@@ -1973,6 +2548,40 @@ LSQUnit::recvRetry()
+void
+LSQUnit::dumpStorePipe()
+{
+    DPRINTF(LSQUnit, "Dumping StorePipe:\n");
```
Same as above

tastynoob commented on src/cpu/o3/lsq_unit.hh:
```diff
@@ -586,6 +655,35 @@ class LSQUnit
+        int size;
+
+        DynInstPtr insts[MaxWidth];
+        std::bitset<LdStFlagNum> flags[MaxWidth];
```
I suggest put the ldsrflags into dyninst

happy-lx replied:
In the current design, where a single instruction may appear in both s0 and s3 of the pipeline (fast replay), I feel storing the state in the TimeBuffer would be easier to manage.

tastynoob commented on src/cpu/o3/lsq_unit.hh:
```diff
@@ -586,6 +655,35 @@ class LSQUnit
+        int size;
+
+        DynInstPtr insts[MaxWidth];
+        std::bitset<LdStFlagNum> flags[MaxWidth];
```
same as above

happy-lx committed:
cpu-o3: refactor lsu related code

+ add LdPipeStages and StPipeStages parameters
+ remove redundant storePipeSx code
+ fix dumpLoadStorePipe

Change-Id: Ie8cb7865c3a53265520f11f016dd467c25a3e2a5
